modules/apps/portal/portal-log4j-test/src/test/java/com/liferay/portal/log4j/Log4jTest.java

```diff
@@ -0,0 +1,464 @@
+		throws IOException {
+
+		UnsyncByteArrayOutputStream unsyncByteArrayOutputStream =
+			new UnsyncByteArrayOutputStream();
+
+		StreamUtil.transfer(inputStream, unsyncByteArrayOutputStream, -1, true);
+
+		return unsyncByteArrayOutputStream.toByteArray();
+	}
+
+	private void _assertLog(String expectedOutput, int invokeLineNumber) {
+		Thread currentThread = Thread.currentThread();
+
+		Matcher matcher = _datePattern.matcher(expectedOutput.substring(0, 23));
+
+		Assert.assertTrue(
+			"Output date format should be yyyy-MM-dd HH:mm:ss.SSS",
+			matcher.matches());
+
+		String content = expectedOutput.substring(23);
+
+		Assert.assertEquals(
+			"Output content should be " + content, content,
+			StringBundler.concat(
+				" INFO  [", currentThread.getName(), "][Log4jTest:",
+				invokeLineNumber, "] Test Message"));
+	}
+
+	private void _assertXmlLog(String expectedOutput, int invokeLineNumber)
+		throws Exception {
+
+		Thread currentThread = Thread.currentThread();
+
+		String[] logMessages = StringUtil.splitLines(_baos.toString());
+
+		String consoleOutput = logMessages[logMessages.length - 1];
+
+		SimpleDateFormat simpleDateFormat = new SimpleDateFormat(
+			"yyyy-MM-dd HH:mm:ss.SSS");
+
+		String date = consoleOutput.substring(0, 23);
+
+		Date logDate = simpleDateFormat.parse(date);
+
+		StringBundler sb = new StringBundler(16);
+
+		sb.append("<log4j:event logger=\"");
+		sb.append(Log4jTest.class.getName());
+		sb.append("\" timestamp=\"");
+		sb.append(logDate.getTime());
+		sb.append("\" level=\"INFO\" thread=\"");
+		sb.append(currentThread.getName());
+		sb.append("\">\r\n");
+		sb.append("<log4j:message><![CDATA[");
+		sb.append("Test Message]]></log4j:message>\r\n");
+		sb.append("<log4j:locationInfo class=\"");
+		sb.append(Log4jTest.class.getName());
+		sb.append("\" method=\"testRollingFileAppender\" file=\"");
+		sb.append("Log4jTest.java\" line=\"");
+		sb.append(invokeLineNumber);
+		sb.append("\"/>\r\n");
+		sb.append("</log4j:event>\r\n\r\n");
+
+		Assert.assertEquals(
+			"LogMessage should be " + expectedOutput, expectedOutput,
+			sb.toString());
+	}
+
+	private static final String[] _NAMES = {
+		"level", "level.off", "level.fatal", "level.error", "level.warn",
+		"level.info", "level.debug", "level.trace"
+	};
+
+	private static final ByteArrayOutputStream _baos =
+		new ByteArrayOutputStream();
+	private static final Pattern _datePattern = Pattern.compile(
+		"\\d\\d\\d\\d-\\d\\d-\\d\\d \\d\\d:\\d\\d:\\d\\d.\\d\\d\\d");
+	private static PrintStream _originalOutputStream;
+	private static PrintStream _printStream;
+	private static File _tempFile;
+	private static final Pattern _textFileNamePattern = Pattern.compile(
+		"liferay.\\d\\d\\d\\d-\\d\\d-\\d\\d.log");
+	private static final Pattern _xmlFileNamePattern = Pattern.compile(
+		"liferay.\\d\\d\\d\\d-\\d\\d-\\d\\d.xml");
+
+	private static class TeePrintStream extends PrintStream {
+
+		public TeePrintStream(
```

Comment:
Please try to change this class:
```
		public TeePrintStream(
			PrintStream printStream, UnsyncStringWriter unsyncStringWriter) {

			_printStream = printStream;
			_unsyncStringWriter = unsyncStringWriter;
		}
```
So you can use the `UnsyncStringWriter` for much easier string handling (it supports `reset()` and you can operate with `StringBundler`).

Basically make your TeePrintStream like https://github.com/liferay/liferay-benchmark-ee/blob/17ad80e52ce0498d90164d06e5a8aae6493707bd/grinder/liferay-grinder/src/main/java/com/liferay/benchmark/tools/ant/RecorderTask.java#L298, just replace `_printStream2` with the `UnsyncStringWriter`.

Reply:
And remember to not close `System.out` in `close()` method.